src/cloudAdapter/cluster/clusterAdapter.ts

```diff
@@ -0,0 +1,173 @@
+
+export class ClusterCloudAdapter implements CloudAdapter {
+    async deploy(
+        input: GenezioCloudInput[],
+        projectConfiguration: ProjectConfiguration,
+        cloudAdapterOptions: CloudAdapterOptions,
+    ): Promise<GenezioCloudOutput> {
+        const stage: string = cloudAdapterOptions.stage || "";
+
+        log.info("Deploying your backend project to genezio infrastructure...");
+
+        // Login to container repository
+        let dockerlogin: SpawnSyncReturns<Buffer> | null = null;
+        const containerRegistryInfo = await getContainerRegistry();
+        const containerRegistryCreds = await getContainerRegistryCredentials();
+        const promisesDeploy = input.map(async (element) => {
+            debugLogger.debug(`Get the registry push data for class name ${element.name}.`);
+
+            if (dockerlogin === null) {
+                spawnSync("docker", ["logout", containerRegistryInfo.repository]);
+
+                dockerlogin = spawnSync("docker", [
+                    "login",
+                    "-u",
+                    containerRegistryInfo.username,
+                    "-p",
+                    containerRegistryCreds.password,
+                    containerRegistryInfo.repository,
+                ]);
+
+                if (dockerlogin.status !== 0) {
+                    throw new Error(
+                        `Error during container repository login. Exit code: ${dockerlogin.status}`,
+                    );
+                }
+            }
+
+            debugLogger.debug(`Upload the content to Harbor registry for class ${element.name}.`);
+
+            // tag image
+            // TODO: discuss with team about timestamp based tagging/versioning of images
+            const timestamp = Date.now().toString();
+            const tag =
+                `${containerRegistryInfo.repository}/${containerRegistryInfo.username}/${projectConfiguration.name}/${element.name}:${timestamp}`.toLowerCase();
+
+            projectConfiguration.classes.find((c) => c.name === element.name)!.options = {
+                ...projectConfiguration.classes.find((c) => c.name === element.name)!.options,
+                timestamp,
+            };
+
+            const dockerTag = spawnSync("docker", [
+                "tag",
+                `${projectConfiguration.name}-${element.name}`.toLowerCase(),
+                tag,
+            ]);
+            if (dockerTag.status !== 0) {
+                throw new Error(`Error during container image tag. Exit code: ${dockerTag.status}`);
+            }
+
+            // push image
+            debugLogger.debug(`Pushing the container image for ${element.name}...`);
+            const dockerPush = spawnSync("docker", ["push", tag]);
+            if (dockerPush.status !== 0) {
+                throw new Error(
+                    `Error during container image push. Exit code: ${dockerPush.status}`,
+                );
```

Comment:
@stefandarius implemented `UserError` [1] which is a nice wrapper that will display error messages to our users. More than that, it will not trigger a sentry error  for us because it's expected behaviour.

Everywhere where you want to show the error message to the user please use `throw new UserError("message")` instead of `Error` - example [2].

[1] https://github.com/Genez-io/genezio/blob/cd9734ed521094d44fd561ffb270f369695d726b/src/errors.ts#L3
[2] https://github.com/Genez-io/genezio/blob/cd9734ed521094d44fd561ffb270f369695d726b/src/requests/getUser.ts#L11